### PR TITLE
Users cannot add an answer to a question that is older than 30 days

### PR DESCRIPTION
**Pull request by rhuberdeau**
Issue #12 

This is my first go at it, but we haven't discussed what the front end will look like. Do we want to remove the new answer form from the page if the question is older than 30 days?

### CODE REVIEW

File: spec/models/answer_spec.rb [hunk 1]
```diff
@@ -14,6 +14,16 @@
     it { should validate_presence_of(:question_id) }
     it { should validate_presence_of(:author_id) }
     it { should validate_length_of(:body).is_at_least(2) }
+
+    describe "#question_not_expired" do 
+      let(:question)  { create :question, created_at: Date.today - 31.days}
```
Comment (samnang) [inline spec/models/answer_spec.rb]: `31.days.ago` should just work as well.

**Comment by samnang** [comment]
From my point of view, I think we either have a flag(expired) on question and have job run daily to update questions that over 30 days or leave as it is and have a method `Question#expired?` in question model.

The advantage of adding a flag to question model, it will be easier to query and faster with partial index.

I totally agree we shouldn't show new answer form instead we might want to show somewhere either text description or badge status to tell to users.

File: app/models/answer.rb [hunk 2]
```diff
@@ -9,6 +9,12 @@ class Answer < ActiveRecord::Base
 
   validates :body, :question_id, :author_id, presence: true
   validates :body, length: { minimum: 2 }
+  validate  :question_not_expired
 
   scope :recent, -> { order('created_at DESC') }
+
+  def question_not_expired
+    errors.add(:question_expired, "The question you were trying to answer was odler than 30 days") if
+      !question_id.blank? and question.created_at < Date.today - 30.days
```
Comment (samnang) [inline app/models/answer.rb]: You can move this logic to `Question#expired?`

**Comment by rhuberdeau** [comment]
We'll still need to add an error to any answer object that comes through- in case someone tries getting past the missing form with a POST request. But I can move the logic to the Question model and query that to see if we'll allow the new answer to be added.

Anyone else have an opinion on the flag/method debate?

**Comment by colinwahl** [comment]
Maybe we could add a can_be_answered? method to the question model.  Then we could check that whenever a user tries to answer a question?  I'll look into implementing it tomorrow, it has been a long week at work so I haven't had much time to look over the source code yet

**Comment by rhuberdeau** [comment]
I've already started working on it. I can commit the updates with the updated tests either tonight or tomorrow.

**Comment by colinwahl** [comment]
Just saw your method. Looks good to me. I'll get started on some other issues tomorrow.

**Comment by rhuberdeau** [comment]
I've committed some updates based on the feedback I got. I decided to hold off on a con job task for now since it could be a pain for people who are hosting on Heroku. Let me know what you guys think.

File: app/models/question.rb [hunk 3]
```diff
@@ -36,14 +36,18 @@ def starred_by?(user)
     evaluation_by(:stars, user) == 1
   end
 
+  def expired?
```
Comment (NARKOZ) [inline app/models/question.rb]: I agree with @samnang that it's better to add a boolean column instead. But let's name it `closed`.

File: app/views/questions/show.html.slim [hunk 4]
```diff
@@ -33,7 +33,7 @@
         | )
       = render partial: 'answers/answer', collection: @question.answers
 
-    - if user_signed_in?
+    - if user_signed_in? && !@question.expired?
```
Comment (NARKOZ) [inline app/views/questions/show.html.slim]: We also need to check it on controller level.

Comment (rhuberdeau) [inline app/views/questions/show.html.slim]: I'm not sure what we can check in the controller action or what we would do with that information. Could you elaborate? 

I  did add a new validation on answers to make sure users cant answer closed questions.

File: app/models/question.rb (hunk 3 above)
Comment (rhuberdeau) [inline app/models/question.rb]: So add a new column 'closed' and run a cron job nightly to update questions that are older than 30 days?

**Comment by rhuberdeau** [comment]
I've made updates to use a boolean value and run a job each night at 2:30 am to update questions that need to be closed.

File: app/models/answer.rb [hunk 5]
```diff
@@ -9,6 +9,12 @@ class Answer < ActiveRecord::Base
 
   validates :body, :question_id, :author_id, presence: true
   validates :body, length: { minimum: 2 }
+  validate  :question_not_closed
 
   scope :recent, -> { order('created_at DESC') }
+
+  def question_not_closed
+    errors.add(:question_closed, "The question you were trying to answer is older than 30 days") if
+      !question_id.blank? and question.closed
```
Comment (NARKOZ) [inline app/models/answer.rb]: Any reason why we check `question_id` presence here?

File: lib/tasks/question_tasks.rake [hunk 6]
```diff
@@ -0,0 +1,7 @@
+desc 'close old questions'
+task close_old_questions: :environment do
+  questions = Question.to_be_closed
+  questions.each do |question|
+    question.update_attributes(closed: true)
+  end
```
Comment (NARKOZ) [inline lib/tasks/question_tasks.rake]: We can rewrite it as `Question.to_be_closed.update_all(closed: true)` to execute only one query.

File: app/views/questions/show.html.slim (hunk 4 above)
Comment (NARKOZ) [inline app/views/questions/show.html.slim]: Sorry, I've missed `question_not_closed` validation in model.

File: app/models/answer.rb (hunk 5 above)
Comment (rhuberdeau) [inline app/models/answer.rb]: Well, I suppose the only way it wouldn't be there is if someone is doing something they shouldn't. I can remove it if you want but I was thinking better safe than sorry.

File: lib/tasks/question_tasks.rake (hunk 6 above)
Comment (rhuberdeau) [inline lib/tasks/question_tasks.rake]: Done.

File: app/models/answer.rb (hunk 5 above)
Comment (NARKOZ) [inline app/models/answer.rb]: We already have validation for `question_id` on line 10, so it's safe to remove it here.

Comment (rhuberdeau) [inline app/models/answer.rb]: It turns out that we do need that IF we want to test for the presence of question_id separately.

**Comment by rhuberdeau** [comment]
I dug through this, one line at a time and this is the only way I could get all the tests to pass.

**Comment by rhuberdeau** [comment]
Any more feedback on this or can it be merged?

**Comment by rhuberdeau** [comment]
Unless anyone has anything else to add I'm going to merge this.

**Pull request merged**